packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -629,3 +656,157 @@ class AsyncGuidelinePainter extends CustomPainter {
+  TimelineGridPainter({
+    @required this.zoom,
+    @required this.constraints,
+    @required this.verticalScrollOffset,
+    @required this.horizontalScrollOffset,
+    @required this.chartStartInset,
+    @required this.chartEndInset,
+    @required this.flameChartWidth,
+    @required this.duration,
+  });
+
+  static const baseGridIntervalPx = 150.0;
+  static const gridLineColor = ThemedColor(
+    Color(0xFFCCCCCC),
+    Color(0xFF585858),
+  );
+  static const timestampOffset = 6.0;
+  static const timestampColor = ThemedColor(
+    Color(0xFF24292E),
+    Color(0xFFFAFBFC),
+  );
+
+  static const origin = 0.0;
+
+  final double zoom;
+
+  final BoxConstraints constraints;
+
+  final double verticalScrollOffset;
+
+  final double horizontalScrollOffset;
+
+  final double chartStartInset;
+
+  final double chartEndInset;
+
+  final double flameChartWidth;
+
+  final Duration duration;
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    // The absolute coordinates of the flame chart's visible section.
+    final visible = Rect.fromLTWH(
+      horizontalScrollOffset,
+      verticalScrollOffset,
+      constraints.maxWidth,
+      constraints.maxHeight,
+    );
+
+    // Paint background for the section that will contain the timestamps. This
+    // section will appear sticky to the top of the viewport.
+    canvas.drawRect(
+      Rect.fromLTWH(
+        origin,
+        origin,
+        constraints.maxWidth,
+        rowHeight,
+      ),
+      Paint()..color = chartBackgroundColor,
+    );
+
+    // Paint the timeline grid lines and corresponding timestamps in the flame
+    // chart.
+    final intervalWidth = _intervalWidth();
+    final microsPerInterval = _microsPerInterval(intervalWidth);
+    int timestampMicros = _startingTimestamp(intervalWidth, microsPerInterval);
+    double lineX;
+    if (visible.left <= chartStartInset) {
+      lineX = chartStartInset - visible.left;
+    } else {
+      lineX =
+          intervalWidth - ((visible.left - chartStartInset) % intervalWidth);
+    }
+
+    while (lineX < constraints.maxWidth) {
+      _paintTimestamp(canvas, timestampMicros, intervalWidth, lineX);
+      _paintGridLine(canvas, lineX);
+      lineX += intervalWidth;
+      timestampMicros += microsPerInterval;
+    }
+  }
+
+  void _paintTimestamp(
+    Canvas canvas,
+    int timestampMicros,
+    double intervalWidth,
+    double lineX,
+  ) {
+    final timestampText = msText(
+      Duration(microseconds: timestampMicros),
+      fractionDigits: timestampMicros == 0 ? 1 : 3,
+    );
+    final textPainter = TextPainter(
+      text: TextSpan(
+        text: timestampText,
+        style: const TextStyle(color: timestampColor),
+      ),
+      textAlign: TextAlign.right,
+      textDirection: TextDirection.ltr,
+    )..layout(maxWidth: intervalWidth);
+
+    // TODO(kenz): figure out a way for the timestamps to scroll out of view
+    // smoothly instead of dropping off.
```

Comment:
if you use a horizontal ListView for the timestamps and use Text widgets you would get that for free. It may be a bit more performant as well as it might do a better job of caching text rendering

Reply:
added a this to the TODO since it may take a bit of refactoring where we calculate intervals and such. It also may be kind of tricky because we would have to inject new widgets into the list when we split each interval into two new intervals (when log2(zoom) increases).